### PR TITLE
Budget summary per committee and budget line

### PR DESCRIPTION
**Pull request by Mr-salto**
re implementation of summary of expenses per committee and budget line

### CODE REVIEW

File: expenses/models.py [hunk 1]
```diff
@@ -216,6 +216,7 @@ def save_user_profile(sender, instance, **kwargs):
     instance.profile.save()
 
 
+
```
Comment (RafDevX) [inline expenses/models.py]: Remove extra line

File: compose.yaml [hunk 2]
```diff
@@ -48,3 +54,5 @@ services:
       # Melvin here since he's treasurer at the time of writing.
       KTH_ID: melvinj
 
+volumes:
+  postgres_data:
```
Comment (RafDevX) [inline compose.yaml]: Files should end with a newline

File: stats/urls.py [hunk 3]
```diff
@@ -6,4 +6,9 @@
     url(r'^$', views.index, name='stats-index'),
     url(r'^summary/$', views.summary, name='stats-summary'),
     url(r'^monthly/(?P<year>\d+)/$', views.monthly, name='stats-breakdown'),
+    url(r'^cost_centres/$', views.cost_centres, name='stats-committees'),
+    url(r'^sec_cost_centres/$', views.sec_cost_centres, name='stats-sec_cost_centre'),
+    url(r'^budget_lines/$', views.budget_lines, name='stats-budget_lines'),
+
+
```
Comment (RafDevX) [inline stats/urls.py]: Remove extra lines

File: stats/views.py [hunk 4]
```diff
@@ -46,29 +46,51 @@ def index(request):
 def summary(request):
     if request.method == "POST":
         body_data = json.loads(request.body)
-        # we tried getting it to work with ids, but it didn't,
-        # so we're using names for now
+        print("body = " , body_data)
```
Comment (RafDevX) [inline stats/views.py]: Remove debug print

File: stats/views.py [hunk 5]
```diff
@@ -46,29 +46,51 @@ def index(request):
...
-                budget_line=body_data['budget_line'],
-                cost_centre=body_data['cost_centre'],
-                secondary_cost_centre=body_data['cost_centre'],
+                cost_centre=cost_centre,
+                expense__expense_date__year=year,
+                secondary_cost_centre=secondary_cost_centre,
+                budget_line = budget_line
             ).all()
-
+
```
Comment (RafDevX) [inline stats/views.py]: Remove extra tab

File: stats/views.py [hunk 6]
```diff
@@ -46,29 +46,51 @@ def index(request):
...
+                secondary_cost_centre=secondary_cost_centre,
+                budget_line = budget_line
             ).all()
-
+        
         sum_amount = 0
+        expense_parts_list = []  # To store each expense part for debugging or detailed view
 
         for expense_part in expense_parts:
+            print(expense_part)  # Print for debugging
```
Comment (RafDevX) [inline stats/views.py]: Remove debug print

File: stats/views.py [hunk 7]
```diff
@@ -80,6 +102,98 @@ def summary(request):
     return Response(status=status.HTTP_404_NOT_FOUND)
 
 
+@csrf_exempt
+def sec_cost_centres(request):
+    if request.method == "POST":
+        body_data = json.loads(request.body)
+        print("body = " , body_data)
```
Comment (RafDevX) [inline stats/views.py]: Remove debug print

File: stats/views.py [hunk 8]
```diff
@@ -80,6 +102,98 @@ def summary(request):
...
+            expense__expense_date__year=year
+        ).all()
+        
+        sec_cost_centres = set()
+
+        for expense_part in expense_parts:
+            sec_cost_centres.add(expense_part.secondary_cost_centre)
+            # Optionally add the expense part details to the response
+    
+        print("sec cost centres = " ,sec_cost_centres)
```
Comment (RafDevX) [inline stats/views.py]: Remove debug print

File: stats/views.py [hunk 9]
```diff
@@ -80,6 +102,98 @@ def summary(request):
...
+        return JsonResponse({
+            'sec_cost_centres': sec_cost_centres_list
+        })
+    return Response(status=status.HTTP_404_NOT_FOUND)
+
+@csrf_exempt
+def budget_lines(request):
+    if request.method == "POST":
+        body_data = json.loads(request.body)
+        print("body = " , body_data)
```
Comment (RafDevX) [inline stats/views.py]: Remove debug print

File: stats/views.py [hunk 10]
```diff
@@ -80,6 +102,98 @@ def summary(request):
...
+
+@require_GET
+def cost_centres(request):
+    """
+    Returns the distinct cost centres (committees) from all expenses.
+    """
+    expense_queryset = models.Expense.objects.all()
+    # expense_queryset is a list of expenses 
+
+    print(expense_queryset)
```
Comment (RafDevX) [inline stats/views.py]: Remove debug print

File: templates/stats/index.html [hunk 11]
```diff
@@ -61,97 +61,130 @@ <h2 v-if="selected == 'Antal cashflows'">Flest Cashflows</h2>
...
+          <td>
+            <!-- Year Dropdown -->
+            <div class="select" style="width: 100%;">
+              <select 
+                v-model="selectedYear" 
+                @change="showDropdownN"
+                style="width: 100%;" 
+                :disabled="expense_parts.length > 1"
+              >
+                <option value="">- Year -</option>
```
Comment (RafDevX) [inline templates/stats/index.html]:
```suggestion
                <option value="">- Välj år -</option>
```

**Review by RafDevX — CHANGES_REQUESTED** [review]
Haven't actually reviewed it properly because I'm not on my computer; this is just the obvious what I noticed while scrolling through without reading

File: compose.yaml [hunk 12]
```diff
@@ -17,6 +17,7 @@ services:
       - 8000:8000
     volumes:
       - ./media:/app/media
+      - ./templates:/app/templates
```
Comment (foodelevator) [inline compose.yaml:20]:
```suggestion
```
I guess it should be written in the readme, but this is already handled if you start it with `docker compose up --watch`

File: compose.yaml [hunk 13]
```diff
@@ -34,11 +35,16 @@ services:
...
+    volumes:
+      - postgres_data:/var/lib/postgresql/data
     healthcheck:
       test: ["CMD", "pg_isready", "-U", "cashflow"]
       interval: 2s
       timeout: 2s
       retries: 3
+    ports:
+      - 5432:5432
+
```
Comment (foodelevator) [inline compose.yaml]:
```suggestion
```

File: stats/views.py [hunk 14]
```diff
@@ -80,6 +101,94 @@ def summary(request):
     return Response(status=status.HTTP_404_NOT_FOUND)
 
 
+@csrf_exempt
+def sec_cost_centres(request):
+    if request.method == "POST":
```
Comment (foodelevator) [inline stats/views.py]: This should be a GET request

File: stats/views.py [hunk 15]
```diff
@@ -80,6 +101,94 @@ def summary(request):
...
+        sec_cost_centres_list = list(sec_cost_centres)
+
+        return JsonResponse({
+            'sec_cost_centres': sec_cost_centres_list
+        })
+    return Response(status=status.HTTP_404_NOT_FOUND)
+
+@csrf_exempt
+def budget_lines(request):
+    if request.method == "POST":
```
Comment (foodelevator) [inline stats/views.py]: Also GET

File: templates/stats/index.html [hunk 16]
```diff
@@ -61,97 +61,130 @@ <h2 v-if="selected == 'Antal cashflows'">Flest Cashflows</h2>
...
+
+
     <table class="card">
       <thead>
         <tr>
-          <th>Resultatställe</th>
-          <th>Sek. resultatställe</th>
-          <th>Budgetpost</th>
           <th>År</th>
+          <th> Nämnder </th>
```
Comment (foodelevator) [inline templates/stats/index.html]:
```suggestion
          <th>Resultatställe</th>
```

File: templates/stats/index.html [hunk 17]
```diff
@@ -61,97 +61,130 @@ <h2 v-if="selected == 'Antal cashflows'">Flest Cashflows</h2>
...
+                @change="costCentreHandler()"
+                style="width: 100%;"
+              >
+                <option value="">- Välj -</option>
+                <option
+                  v-for="(costCentre, index) in costCentres"
+                  :key="index"
+                  :value="costCentre"
+                >
+                {% verbatim %}
```
Comment (foodelevator) [inline templates/stats/index.html]: This works, but there is also the `v-text` property

File: compose.yaml [hunk 18]
```diff
@@ -34,11 +35,16 @@ services:
       POSTGRES_USER: cashflow
       POSTGRES_PASSWORD: cashflow
       POSTGRES_DB: cashflow
+    volumes:
+      - postgres_data:/var/lib/postgresql/data
```
Comment (foodelevator) [inline compose.yaml]: I don't see a reason to add this as data is persisted as long as you don't run `docker compose down` and it hides the data for those who already have something in their local databases

File: stats/views.py (hunk 14 above)
Comment (Mr-salto) [reply stats/views.py]: I need to make a POST as every secondary cost center is dependent on the primary cost center. Maybe there is an easier way to do it with a GET.

File: stats/views.py (hunk 15 above)
Comment (Mr-salto) [reply stats/views.py]: Same reason

File: templates/stats/index.html (hunk 17 above)
Comment (Mr-salto) [reply templates/stats/index.html]: Didn't know, thank you !

File: compose.yaml (hunk 18 above)
Comment (Mr-salto) [reply compose.yaml]: I didn't know and wanted to be sure that the fake data I inserted for tests purposes would be persistent.

File: stats/views.py (hunk 14 above)
Comment (foodelevator) [reply stats/views.py]: You can for instance use query parameters to send along data, if that's what you mean

File: compose.yaml [hunk 19]
```diff
@@ -47,4 +50,4 @@ services:
       # Since we're using the real pls, it's nice to use some user with a lot of privileges. Putting
       # Melvin here since he's treasurer at the time of writing.
       KTH_ID: melvinj
-
+
```
Comment (RafDevX) [inline compose.yaml]:
```suggestion

```

File: stats/views.py [hunk 20]
```diff
@@ -42,44 +42,145 @@ def index(request):
...
     })
 
-@csrf_exempt
+@require_GET
 def summary(request):
-    if request.method == "POST":
-        body_data = json.loads(request.body)
-        # we tried getting it to work with ids, but it didn't,
-        # so we're using names for now
+    if request.method == "GET":
```
Comment (RafDevX) [inline stats/views.py]: Aren't these two lines synonymous?

File: templates/stats/index.html [hunk 21]
```diff
@@ -167,77 +188,79 @@ <h2>Antal inköp per månad</h2>
...
-             let secondary_cost_centres = this.committees
-                                    .find(x => x.id == this.selectedCommittee.committeeId)
-                                    .secondary_cost_centres;
-             secondary_cost_centres = secondary_cost_centres.filter(cc => cc.budget_lines.length != 0);
-             return secondary_cost_centres;
-         },
-         activeCommittees: function() {
-             return this.committees.filter(c => !c.inactive);
-         },
+      el: '#content',
```
Comment (RafDevX) [inline templates/stats/index.html]: (can't believe I'm saying this) go back to 5 space indentation; easier for diff

**Review by RafDevX — CHANGES_REQUESTED** [review]

File: templates/stats/index.html [hunk 22]
```diff
@@ -167,77 +188,79 @@ <h2>Antal inköp per månad</h2>
...
+             costCentres: [] ,
+             secondaryCostCentres: [] ,
+             budgetLines: [] ,
+             expense_parts: [ {
+                 costCentre: '',
+                 secondaryCostCentre: '',
+                 budgetLine: '',
+                 amount: 0,
+             }],
+             totalSum: null,
```
Comment (RafDevX) [inline templates/stats/index.html:215]: Confirm whether this is still needed

File: stats/views.py [hunk 23]
```diff
@@ -42,42 +42,136 @@ def index(request):
...
 
-@csrf_exempt
+@require_GET
 def summary(request):
-    if request.method == "POST":
-        body_data = json.loads(request.body)
-        # we tried getting it to work with ids, but it didn't,
-        # so we're using names for now
-
         expense_parts = None
```
Comment (RafDevX) [inline stats/views.py]: Fix indentation in all of these

**Review by RafDevX — CHANGES_REQUESTED** [review]

**Review by RafDevX — APPROVED** [review]

**Pull request merged**